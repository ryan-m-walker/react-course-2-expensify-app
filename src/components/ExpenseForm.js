import React, { Component } from 'react'
import moment from 'moment'
import 'react-dates/initialize'
import { SingleDatePicker } from 'react-dates'


class ExpenseForm extends Component {
  constructor(props) {
    super(props)
    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calenderFocused: false,
      error: false
    }
  }

  onDescriptionChange = e => {
    const description = e.target.value
    this.setState(() => ({ description }))
  }

  onNoteChange = e => {
    const note = e.target.value
    this.setState(() => ({ note }))
  }

  onAmountChange = e => {
    const amount = e.target.value
  if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }))
    }
  }

  onDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({ createdAt }))
    } 
  }

  onFocusChange = ({focused}) => {
    this.setState(() => ({ calenderFocused: focused }))
  }

  onSubmit = e => {
    e.preventDefault()
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please provide a description and amount.' }))
    } else {
      const { description, note, amount, createdAt } = this.state
      this.setState(() => ({ error: '' }))
      this.props.onSubmit({
        description,
        amount: parseFloat(amount, 10) * 100,
        createdAt: createdAt.valueOf(),
        note
      })
    }
  }

  render() {
    const { 
      description, 
      note, 
      amount, 
      createdAt, 
      calenderFocused,
      error,
    } = this.state
    const { 
      onDescriptionChange, 
      onNoteChange, 
      onAmountChange, 
      onDateChange, 
      onFocusChange, 
      onSubmit,
    } = this

    return (
      <div>
        { error && <p>{ error }</p> }
        <form onSubmit={onSubmit}>
          <input
            autoFocus
            placeholder='Description'
            type='text'
            value={description}
            onChange={onDescriptionChange} />
          <input 
            type='text'
            placeholder='Amount'
            onChange={onAmountChange}
            value={amount}/>
          <SingleDatePicker 
            date={createdAt}
            onDateChange={onDateChange}
            focused={calenderFocused}
            onFocusChange={onFocusChange}
            numberOfMonths={1}
            isOutsideRange={day => false} />
          <textarea
            onChange={onNoteChange}
            value={note}
            placeholder='Add a note for your expense (optional)'>
          </textarea>
          <button>
            { this.props.expense ? 'Save Expense' : 'Add Expense' }
          </button>
        </form>
      </div>
    )
  }
}


export default ExpenseForm
