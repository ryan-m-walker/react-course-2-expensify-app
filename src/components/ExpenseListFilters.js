import React, { Component } from 'react'
import { connect } from 'react-redux'
import { DateRangePicker } from 'react-dates';

import { 
  setTextFilter, 
  sortByDate, 
  sortByAmount,
  setStartDate,
  setEndDate
} from '../actions/filters'


export class ExpenseListFilters extends Component {
  state = {
    calenderFocused: null
  }
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate)
    this.props.setEndDate(endDate)
  }
  onFocusChange = calenderFocused => {
    this.setState(() => ({ calenderFocused }))
  }
  onTextChange = e => {
    this.props.setTextFilter(e.target.value)
  }
  onSortChange = e => {
    if (e.target.value === 'date') {
      this.props.sortByDate()
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount()
    }
  }
  render() {
    const { filters, dispatch } = this.props
    const { calenderFocused } = this.state
    const { onTextChange, onSortChange } = this
    return (
      <div>
        <input
          onChange={onTextChange}
          type='text'
          value={filters.text}/>
        <select 
          value={filters.sortBy}
          onChange={onSortChange}>
          <option value='date'>Date</option>
          <option value='amount'>Amount</option>
        </select>
        <DateRangePicker 
          showClearDates
          startDate={filters.startDate}
          endDate={filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={calenderFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false} />
      </div>
    )
  }
}
  


const mapStateToProps = ({ filters }) => ({ filters })

const mapDispatchToProps = dispatch => ({
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)
