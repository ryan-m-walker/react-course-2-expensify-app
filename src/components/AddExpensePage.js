import React, { Component } from 'react'
import { connect } from 'react-redux'

import ExpenseForm from './ExpenseForm'
import { addExpense } from '../actions/expenses'


export class AddExpensePage extends Component {
  onSubmit = expense => {
    this.props.addExpense(expense)
    this.props.history.push('/')
  }
  render() {
    const { history } = this.props
    const { onSubmit } = this
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm 
          onSubmit={onSubmit} />
      </div>
    )
  }
}


const mapDispatchToProps = dispatch => ({
    addExpense: expense => dispatch(addExpense(expense))
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage)
