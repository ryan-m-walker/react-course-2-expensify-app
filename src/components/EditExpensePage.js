import React, { Component } from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'

import { editExpense, removeExpense } from '../actions/expenses' 


export class EditExpensePage extends Component {
  onSubmit = item => {
    const { expense, editExpense, history } = this.props
    editExpense(expense.id, item)
    history.push('/')
  }
  handleRemove = () => {
    const { expense, removeExpense, history } = this.props
    removeExpense({ id: expense.id })
    history.push('/')
  }
  render() {
    const { expense, dispatch, history } = this.props
    const { onSubmit, handleRemove } = this
    return (
      <div>
        <ExpenseForm
          expense={expense} 
          onSubmit={onSubmit} />
        <button
          onClick={handleRemove}>
          Remove
        </button>
      </div>
    )
  }
}


const mapStateToProps = (state, props) => ({
    expense: state.expenses.filter(expense => 
      expense.id === props.match.params.id
    )[0]
})

const mapDispatchToProps = dispatch => ({
  editExpense: (id, item) => dispatch(editExpense(id, item)),
  removeExpense: id => dispatch(removeExpense(id))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)
