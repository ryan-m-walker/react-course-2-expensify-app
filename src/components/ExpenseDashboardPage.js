import React from 'react'
import { connect } from 'react-redux'

import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'
import selectExpenses from '../selectors/expenses'
import getExpensesTotal from '../selectors/expenses-total'
import ExpensesSummary from './ExpensesSummary'


export const ExpenseDashboardPage = ({expenses}) => (
  <div>
    <ExpenseListFilters />
    <ExpensesSummary />
    <ExpenseList />
  </div>
)



const mapState = state => ({
  expenses: selectExpenses(state.expenses, state.filters)
})

export default connect(mapState)(ExpenseDashboardPage)
