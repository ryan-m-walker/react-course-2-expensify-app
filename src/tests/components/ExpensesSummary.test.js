import React from 'react'
import { shallow } from 'enzyme'

import { ExpensesSummary } from '../../components/ExpensesSummary' 

test('Should correctly render ExpensesSummary with 1 expense', () => {
  const wrapper = shallow(
    <ExpensesSummary 
      expenseCount={1}
      expensesTotal={1231} />
  )
  expect(wrapper).toMatchSnapshot()

})

test('Should correctly render ExpensesSummary with multiple expense', () => {
  const wrapper = shallow(
    <ExpensesSummary 
      expenseCount={23}
      expensesTotal={1231123} />
  )
  expect(wrapper).toMatchSnapshot()
})
  