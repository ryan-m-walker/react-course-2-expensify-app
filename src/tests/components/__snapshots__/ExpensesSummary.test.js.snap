// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should correctly render ExpensesSummary with 1 expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $12.31
  </h1>
</div>
`;

exports[`Should correctly render ExpensesSummary with multiple expense 1`] = `
<div>
  <h1>
    Viewing 
    23
     
    expenses
     totalling 
    $12,311.23
  </h1>
</div>
`;
